### PR TITLE
vision position covariance usage

### PR DESCRIPTION
**Pull request by ChristophTobler**
I added the validation parameters in the `vehicle_vision_position` to signalize the EKF2 if it should use it. Currently I use the `eph/epv` parameters for passing through noise parameters and a threshold of `100` to set it valid/invalid.
@priseborough thoughts?

### CODE REVIEW

**Comment by ChristophTobler** [comment]
I guess long term it makes sense to have a separate `vehicle_vision_position` msg (currently it's a multi-topic from `vehicle_local_position`) that includes the covariance matrix (upper right half) and some other parameters as `reset`, `loop_closure_detected` etc.

File: src/modules/ekf2/ekf2_main.cpp [hunk 1]
```diff
@@ -858,11 +858,14 @@ void Ekf2::run()
 			ev_data.quat = q;
 
 			// position measurement error from parameters. TODO : use covariances from topic
-			ev_data.posErr = _ev_pos_noise.get();
+			ev_data.posErr = fmaxf(_ev_pos_noise.get(), fmaxf(ev_pos.eph, ev_pos.epv));
```
Comment (priseborough) [inline src/modules/ekf2/ekf2_main.cpp]: Longer term  I will update the ecl EKF to handle separate horiz and vertical accuracy values for ev data.

**Comment by priseborough** [comment]
I've checked the ecl EKF code and the result of stopping EV messages should be that after 5 seconds a "EKF External Vision Data Stopped" information message is sent to the console. If EV data is the only aiding source, then the EKF should fall back to a attitude and vertical pos/vel estimation only mode together with a "EKF stopping navigation" warning message.

This needs to be verified with testing before merging these changes.

File: src/modules/ekf2/ekf2_main.cpp [hunk 2]
```diff
@@ -858,11 +858,14 @@ void Ekf2::run()
...
 			ev_data.angErr = _ev_ang_noise.get();
 
-			// use timestamp from external computer, clocks are synchronized when using MAVROS
-			_ekf.setExtVisionData(vision_position_updated ? ev_pos.timestamp : ev_att.timestamp, &ev_data);
+			// only set data if all positions and velocities are valid
+			if (ev_pos.xy_valid && ev_pos.z_valid && ev_pos.v_xy_valid && ev_pos.v_z_valid) {
```
Comment (priseborough) [inline src/modules/ekf2/ekf2_main.cpp]: I've checked the ecl EKF code and the result of stopping EV messages should be that after 5 seconds a "EKF External Vision Data Stopped" information message is sent to the console. If EV data is the only aiding source, then the EKF should fall back to a attitude and vertical pos/vel estimation only mode together with a "EKF stopping navigation" warning message.

This needs to be verified with testing before merging these changes.

Comment (ChristophTobler) [reply src/modules/ekf2/ekf2_main.cpp]: I'll test it in simulation and will share the logs. Thanks!

**Comment by priseborough** [comment]
I've checked the ecl EKF code and the result of stopping EV messages should be that after 5 seconds a "EKF External Vision Data Stopped" information message is sent to the console. If EV data is the only aiding source, then the EKF should fall back to a attitude and vertical pos/vel estimation only mode together with a "EKF stopping navigation" warning message.

This needs to be verified with testing before merging these changes because the combined effect of stopping a data source that is providing position, height and yaw reference needs to be understood.

**Comment by ChristophTobler** [comment]
Here's a log https://review.px4.io/plot_app?log=e095c33e-e72d-48a1-aa5e-8386581fe46a
It switched back to altitude mode but then always did `EKF baro hgt timeout - reset to baro` and gained altitude.

**Comment by TSC21** [comment]
@ChristophTobler can you please check if this does not break anything on LPE side? We actually use the covariance values there: https://github.com/PX4/Firmware/blob/master/src/modules/local_position_estimator/sensors/vision.cpp#L86-L101. The way I see it it won't break, cause we are relaying on the eph and epv values, but better make sure first as we are actually changing status values. Maybe you can add the status verification there also as you are doing with the EKF2? Thanks

**Comment by ChristophTobler** [comment]
@priseborough I had to add some changes in ECL (see https://github.com/PX4/ecl/pull/379). Could you please verify? Log: https://review.px4.io/plot_app?log=f2d565a4-d0a6-4806-9345-27a84a549d90
@TSC21 This has no effect on LPE as it is right now. LPE will just continue using EV with a large standard deviation. https://review.px4.io/plot_app?log=92934d8e-384a-4b99-be66-86c13c5b2ed1
We can add the same checks to the LPE but I'm not familiar with the LPE. If I just add a check e.g. [here](https://github.com/PX4/Firmware/blob/master/src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp#L504), it will "fall" from the sky with `LPE_FUSION` set to `28` or `156` (plus baro) Log: https://review.px4.io/plot_app?log=a042f9c8-83a9-4192-b5b4-a5d1e8cf457d.
So I would probably separate a LPE PR.

**Comment by priseborough** [comment]
@ChristophTobler Behaviour in log looks OK as do ecl code changes. I've merged the ecl changes.

**Comment by ChristophTobler** [comment]
@priseborough Thanks! I've updated the ecl submodule

**Pull request merged**